### DISCUSSION TITLE
Debugging Air Project when not at the root folder

### DISCUSSION THREAD

**Discussion by markdemich** [BowlerHatLLC/vscode-as3mxml#671] [Q&A] [answered]
I have a repo where it has multiple folders.  Most of the folders represent a different project where I build flex modules (swc).  I also have a folder with an Air application that ultimately uses those modules.  When I tried to debug, I could not figure out how to get the debugger or launch.json to recognized the app in the folder.  Instead, I needed to open that folder separately inside a new VSCode window in order to debug.  It's a decent work around, but not idea.  Is there a way to get this to work otherwise?

**Comment by joshtynjala** [comment, marked as answer]
It is possible to debug when you open a common parent folder as the root of a workspace, but you'll need to configure your _.vscode/launch.json_ file manually. I think the only thing you are required to do differently is set the `program` field:

```json
{
	"type": "swf",
	"request": "launch",
	"name": "Launch SWF",
	"program": "appfolder/bin-debug/MyProject-app.xml"
}
```

See [this page](https://github.com/BowlerHatLLC/vscode-as3mxml/wiki/launch.json-fields-for-SWF-debugging) for all available fields of the swf debugger.

There is another way, though, and I would even argue that this approach is better. Right now, you're opening the common parent folder that contains all of your individual projects. Instead, start  by opening your app project folder directly. Then use VSCode's _Add Folder to workspace_ for each individual library project. Then, use VSCode's _Save workspace as_ to create a _.code-workspace_ file. Each project will be a separate root folder, and you can create _.vscode/launch.json_ directly in the app project, and it will be auto-configured from your _asconfig.json_ without setting any fields manually.

**Reply by joshtynjala** [reply]
For reference, this is known as a [multi-root workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces).

**Reply by joshtynjala** [reply]
Although, if you decide to go with a multi-root workspace, I don't think it will be compatible with [your use of `dependsOn` in tasks.json](https://github.com/BowlerHatLLC/vscode-as3mxml/discussions/670#discussioncomment-5079295), since the individual workspaces won't know about each other's tasks. It's an unfortunate tradeoff.

I prefer multi-root workspaces, so for one of my apps, I just added the libraries to the app project's `source-path` instead of compiling them to SWC first. Since it would have needed to recompile the library SWCs in VSCode every time anyway, it is actually a bit faster to launch the compiler only once.

**Comment by markdemich** [comment]
Ok, thanks so much.  Looks like I have some options to choose from.  I
appreciate you time in writing this all out for me.
